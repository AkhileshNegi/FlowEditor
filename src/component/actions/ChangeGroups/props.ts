import { ChangeGroups } from '../../../flowTypes';
import { Type } from '../../../config';

export default interface ChangeGroupsFormProps {
    typeConfig: Type;
    action: ChangeGroups;
    updateAction: (action: ChangeGroups) => void;
    onBindWidget: (ref: any) => void;
    removeWidget: (name: string) => void;
};
