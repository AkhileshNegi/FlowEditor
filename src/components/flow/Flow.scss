@import '~/variables.scss';

.loading {
    visibility: hidden;
    height: inherit;
}

.loaded {
    height: inherit;
    position: relative;
}

.nodeList {
    position: relative;
    user-select: none;
}

.dragSelection {
    position: absolute;
    background: red;
    border: 0px solid darken($selection_square, 10%);
    opacity: 0;
    pointer-events: none;
    z-index: $z_selection;
}
