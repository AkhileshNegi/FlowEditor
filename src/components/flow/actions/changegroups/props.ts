import { ActionFormHelper, Type } from '~/config/typeConfigs';
import { ChangeGroups } from '~/flowTypes';
import { NodeEditorSettings } from '~/store/nodeEditor';

export interface ChangeGroupsFormProps {
    // action details
    nodeSettings: NodeEditorSettings;
    formHelper: ActionFormHelper;
    typeConfig: Type;

    // update handlers
    updateAction(action: ChangeGroups): void;

    // modal notifiers
    onTypeChange(config: Type): void;
    onClose(canceled: boolean): void;
}
export default ChangeGroupsFormProps;
