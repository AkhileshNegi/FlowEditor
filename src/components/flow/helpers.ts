import { Type, Types } from '~/config/interfaces';
import { getType, getTypeConfig } from '~/config/typeConfigs';
import { NodeEditorSettings } from '~/store/nodeEditor';

export const determineTypeConfig = (nodeSettings: NodeEditorSettings): Type => {
    const renderNode = nodeSettings.originalNode;
    const node = renderNode && renderNode.node;

    if (nodeSettings.originalAction && nodeSettings.originalAction.type) {
        return getTypeConfig(nodeSettings.originalAction.type);
    } else if (node && node.actions && node.actions.length > 0) {
        return getTypeConfig(node.actions[node.actions.length - 1].type);
    } else {
        try {
            const type = getType(renderNode);
            const config = getTypeConfig(type);
            if (config.type !== Types.missing) {
                return config;
            }
            // tslint:disable-next-line:no-empty
        } catch (Error) {}
    }

    // Account for ghost nodes
    if (node && node.router) {
        return getTypeConfig(node.router.type);
    }

    throw new Error(`Couldn't determine type config for: ${node.uuid}`);
};
