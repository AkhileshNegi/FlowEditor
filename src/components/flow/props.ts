import { UpdateLocalizations } from '~/components/nodeeditor/NodeEditor';
import { Type } from '~/config/typeConfigs';
import { AnyAction } from '~/flowTypes';
import { Asset } from '~/services/AssetService';
import { AssetStore, RenderNode } from '~/store/flowContext';
import { NodeEditorSettings } from '~/store/nodeEditor';
import { DispatchWithState, GetState } from '~/store/thunks';

export interface ActionFormProps {
    // action details
    nodeSettings: NodeEditorSettings;
    typeConfig: Type;
    assets: AssetStore;

    // update handlers
    updateAction(
        action: AnyAction,
        onUpdated?: (dispatch: DispatchWithState, getState: GetState) => void
    ): void;

    // modal notifiers
    onTypeChange(config: Type): void;
    onClose(canceled: boolean): void;
}

export interface RouterFormProps {
    nodeSettings: NodeEditorSettings;
    typeConfig: Type;

    assets: AssetStore;

    // update handlers
    updateRouter(renderNode: RenderNode): void;

    // modal notifiers
    onTypeChange(config: Type): void;
    onClose(canceled: boolean): void;
}

export interface LocalizationFormProps {
    language: Asset;
    nodeSettings: NodeEditorSettings;
    updateLocalizations(languageCode: string, localizations: any[]): UpdateLocalizations;
    onClose(canceled: boolean): void;
}
