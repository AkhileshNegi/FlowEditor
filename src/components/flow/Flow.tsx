import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as styles from '~/components/flow/Flow.scss';
import ConnectedNode from '~/components/flow/node/Node';
import { getDraggedFrom } from '~/components/helpers';
import ConnectedNodeEditor from '~/components/nodeeditor/NodeEditor';
import Simulator from '~/components/simulator/Simulator';
import Sticky from '~/components/sticky/Sticky';
import { ConfigProviderContext } from '~/config';
import { fakePropType } from '~/config/ConfigProvider';
import { getActivity } from '~/external';
import { FlowDefinition, FlowPosition } from '~/flowTypes';
import ActivityManager from '~/services/ActivityManager';
import Plumber from '~/services/Plumber';
import { DragSelection, EditorState } from '~/store/editor';
import { RenderNode } from '~/store/flowContext';
import { getCollisions } from '~/store/helpers';
import { NodeEditorSettings } from '~/store/nodeEditor';
import AppState from '~/store/state';
import {
    ConnectionEvent,
    DispatchWithState,
    ensureStartNode,
    EnsureStartNode,
    mergeEditorState,
    MergeEditorState,
    NoParamsAC,
    OnConnectionDrag,
    onConnectionDrag,
    OnDragSelection,
    onDragSelection,
    OnOpenNodeEditor,
    onOpenNodeEditor,
    resetNodeEditingState,
    UpdateConnection,
    updateConnection,
    updateSticky,
    UpdateSticky
} from '~/store/thunks';
import { contextTypes } from '~/testUtils';
import {
    createUUID,
    isRealValue,
    NODE_PADDING,
    renderIf,
    snapToGrid,
    timeEnd,
    timeStart
} from '~/utils';
import Debug from '~/utils/debug';

export interface FlowStoreProps {
    editorState: Partial<EditorState>;
    mergeEditorState: MergeEditorState;

    definition: FlowDefinition;
    nodes: { [uuid: string]: RenderNode };
    dependencies: FlowDefinition[];
    nodeEditorSettings: NodeEditorSettings;

    ensureStartNode: EnsureStartNode;
    updateConnection: UpdateConnection;
    onOpenNodeEditor: OnOpenNodeEditor;
    onDragSelection: OnDragSelection;
    resetNodeEditingState: NoParamsAC;
    onConnectionDrag: OnConnectionDrag;
    updateSticky: UpdateSticky;
}

export interface Translations {
    [uuid: string]: any;
}

export const DRAG_THRESHOLD = 8;
export const REPAINT_TIMEOUT = 500;
export const GHOST_POSITION_INITIAL = { left: -1000, top: -1000 };

export const nodeSpecId = 'node';
export const nodesContainerSpecId = 'node-container';
export const ghostNodeSpecId = 'ghost-node';
export const dragSelectSpecId = 'drag-select';

export const isDraggingBack = (event: ConnectionEvent) => {
    return event.suspendedElementId === event.targetId && event.source !== null;
};

export const getDragStyle = (drag: DragSelection) => {
    const left = Math.min(drag.startX, drag.currentX);
    const top = Math.min(drag.startY, drag.currentY);
    const width = Math.max(drag.startX, drag.currentX) - left;
    const height = Math.max(drag.startY, drag.currentY) - top;
    return {
        left,
        top,
        width,
        height
    };
};

export class Flow extends React.Component<FlowStoreProps, {}> {
    private Activity: ActivityManager;
    private Plumber: Plumber;
    private containerOffset = { left: 0, top: 0 };
    private ele: HTMLDivElement;
    private nodeContainerUUID: string;

    // Refs
    private ghost: any;

    public static contextTypes = {
        endpoints: fakePropType,
        debug: fakePropType
    };

    constructor(props: FlowStoreProps, context: ConfigProviderContext) {
        super(props, context);

        this.nodeContainerUUID = createUUID();

        this.Activity = new ActivityManager(this.props.definition.uuid, getActivity);

        this.Plumber = new Plumber();

        /* istanbul ignore next */
        if (context.debug) {
            window.fe = new Debug(props, this.props.editorState.debug);
        }

        bindCallbacks(this, {
            include: [/Ref$/, /^on/, /^is/]
        });

        timeStart('RenderAndPlumb');
    }

    private onRef(ref: HTMLDivElement): HTMLDivElement {
        return (this.ele = ref);
    }

    private ghostRef(ref: any): any {
        return (this.ghost = ref);
    }

    public componentDidMount(): void {
        this.Plumber.bind('connection', (event: ConnectionEvent) =>
            this.props.updateConnection(event.sourceId, event.targetId)
        );
        this.Plumber.bind('beforeDrag', (event: ConnectionEvent) =>
            this.beforeConnectionDrag(event)
        );

        this.Plumber.bind('connectionDrag', (event: ConnectionEvent) => {
            this.props.onConnectionDrag(event);
        });

        this.Plumber.bind('connectionDragStop', (event: ConnectionEvent) =>
            this.onConnectorDrop(event)
        );
        this.Plumber.bind(
            'beforeStartDetach',
            (event: ConnectionEvent) => !this.props.editorState.translating
        );
        this.Plumber.bind('beforeDetach', (event: ConnectionEvent) => true);
        this.Plumber.bind('beforeDrop', (event: ConnectionEvent) =>
            this.onBeforeConnectorDrop(event)
        );

        // If we don't have any nodes, create our first one
        this.props.ensureStartNode();

        let offset = { left: 0, top: 0 };

        /* istanbul ignore next */
        if (this.ele) {
            offset = this.ele.getBoundingClientRect();
        }
        this.containerOffset = { left: offset.left, top: offset.top + window.scrollY };

        timeEnd('RenderAndPlumb');

        // deals with safari load rendering throwing
        // off the jsplumb offsets
        window.setTimeout(() => this.Plumber.repaint(), REPAINT_TIMEOUT);
    }

    public componentWillUnmount(): void {
        this.Plumber.reset();
    }

    /**
     * Called right before a connector is dropped onto a new node
     */
    private onBeforeConnectorDrop(event: ConnectionEvent): boolean {
        this.props.resetNodeEditingState();

        if (event.sourceId.split(':')[0] === event.targetId) {
            console.error(event.targetId + ' cannot point to itself');
            return false;
        }

        return true;
    }

    // private onShowDefinition(definition: FlowDefinition): void {
    //     TODO: make this work, it's cool!
    //     this.Plumber.reset();
    //     this.setState({ viewDefinition: definition }, () => { this.Plumber.repaint() });
    // }

    /**
     * Called the moment a connector is done dragging, whether it is dropped on an
     * existing node or on to empty space.
     */
    private onConnectorDrop(event: ConnectionEvent): boolean {
        const { ghostNode } = this.props.editorState;
        // Don't show the node editor if we a dragging back to where we were
        if (isRealValue(ghostNode) && !isDraggingBack(event)) {
            // Wire up the drag from to our ghost node
            this.Plumber.recalculate(ghostNode.node.uuid);

            const dragPoint = getDraggedFrom(ghostNode);
            this.Plumber.connect(
                dragPoint.nodeUUID + ':' + dragPoint.exitUUID,
                ghostNode.node.uuid
            );

            // Save our position for later
            const { left, top } = (this.ghost &&
                snapToGrid(
                    this.ghost.wrappedInstance.ele.offsetLeft,
                    this.ghost.wrappedInstance.ele.offsetTop
                )) || { left: 0, top: 0 };

            this.props.editorState.ghostNode.ui.position = { left, top };
            let originalAction = null;
            if (ghostNode.node.actions && ghostNode.node.actions.length === 1) {
                originalAction = ghostNode.node.actions[0];
            }

            // Bring up the node editor
            this.props.onOpenNodeEditor({
                originalNode: ghostNode,
                originalAction
            });
        }

        // To-do: mock this out
        /* istanbul ignore next */
        $(document).off('mousemove');

        return true;
    }

    private beforeConnectionDrag(event: ConnectionEvent): boolean {
        return !this.props.editorState.translating;
    }

    private getNodes(): JSX.Element[] {
        return Object.keys(this.props.nodes).map(uuid => {
            const renderNode = this.props.nodes[uuid];
            return (
                <div
                    key={uuid}
                    /*{...createClickHandler(() => {
                        console.log('do it');
                    })}*/

                    onMouseDown={(event: React.MouseEvent<HTMLDivElement>) => {
                        // this.props.mergeEditorState({});
                        this.onDragNodeStart(renderNode, {
                            left:
                                event.pageX -
                                this.containerOffset.left -
                                renderNode.ui.position.left,
                            top: event.pageY - this.containerOffset.top - renderNode.ui.position.top
                        });
                    }}
                    onMouseUp={(event: React.MouseEvent<HTMLDivElement>) => {
                        this.onDragNodeStop(renderNode.node.uuid);
                    }}
                >
                    <ConnectedNode
                        key={renderNode.node.uuid}
                        data-spec={nodeSpecId}
                        nodeUUID={renderNode.node.uuid}
                        Activity={this.Activity}
                        plumberRepaintForDuration={this.Plumber.repaintForDuration}
                        plumberMakeTarget={this.Plumber.makeTarget}
                        plumberRemove={this.Plumber.remove}
                        plumberRecalculate={this.Plumber.recalculate}
                        plumberMakeSource={this.Plumber.makeSource}
                        plumberConnectExit={this.Plumber.connectExit}
                        plumberUpdateClass={this.Plumber.updateClass}
                    />
                </div>
            );
        });
    }

    private getStickies(): JSX.Element[] {
        const stickyMap = this.props.definition._ui.stickies || {};
        return Object.keys(stickyMap).map(uuid => {
            return <Sticky key={uuid} uuid={uuid} sticky={stickyMap[uuid]} />;
        });
    }

    private getDragNode(): JSX.Element {
        return isRealValue(this.props.editorState.ghostNode) ? (
            <ConnectedNode
                ref={this.ghostRef}
                key={this.props.editorState.ghostNode.node.uuid}
                data-spec={ghostNodeSpecId}
                ghost={true}
                nodeUUID={this.props.editorState.ghostNode.node.uuid}
                Activity={this.Activity}
                plumberRepaintForDuration={this.Plumber.repaintForDuration}
                plumberMakeTarget={this.Plumber.makeTarget}
                plumberRemove={this.Plumber.remove}
                plumberRecalculate={this.Plumber.recalculate}
                plumberMakeSource={this.Plumber.makeSource}
                plumberConnectExit={this.Plumber.connectExit}
                plumberUpdateClass={this.Plumber.updateClass}
            />
        ) : null;
    }

    private getSimulator(): JSX.Element {
        return renderIf(this.context.endpoints && this.context.endpoints.simulateStart)(
            <Simulator Activity={this.Activity} />
        );
    }

    private getNodeEditor(): JSX.Element {
        return renderIf(this.props.nodeEditorSettings !== null)(
            <ConnectedNodeEditor
                plumberConnectExit={this.Plumber.connectExit}
                plumberRepaintForDuration={this.Plumber.repaintForDuration}
            />
        );
    }

    private isClickOnCanvas(event: React.MouseEvent<HTMLDivElement>): boolean {
        // TODO: not sure the TS-safe way to access id here
        return (event.target as any).id === this.nodeContainerUUID;
    }

    private onDoubleClick(event: React.MouseEvent<HTMLDivElement>): void {
        if (this.isClickOnCanvas(event)) {
            const { left, top } = snapToGrid(
                event.clientX - this.containerOffset.left - 100 + NODE_PADDING,
                event.clientY - this.containerOffset.top - NODE_PADDING * 2
            );

            this.props.updateSticky(createUUID(), {
                position: { left, top },
                title: 'New Note',
                body: '...'
            });
        }
    }

    public onMouseDown(event: React.MouseEvent<HTMLDivElement>): void {
        if (this.isClickOnCanvas(event)) {
            this.props.mergeEditorState({
                dragSelection: {
                    startX: event.pageX - this.containerOffset.left,
                    startY: event.pageY - this.containerOffset.top,
                    currentX: event.pageX - this.containerOffset.left,
                    currentY: event.pageY - this.containerOffset.top,
                    selected: null
                }
            });
        }
    }

    public onMouseMove(event: React.MouseEvent<HTMLDivElement>): void {
        if (this.props.editorState.dragSelection && this.props.editorState.dragSelection.startX) {
            const drag = this.props.editorState.dragSelection;
            const left = Math.min(drag.startX, drag.currentX);
            const top = Math.min(drag.startY, drag.currentY);
            const right = Math.max(drag.startX, drag.currentX);
            const bottom = Math.max(drag.startY, drag.currentY);

            this.props.mergeEditorState({
                dragSelection: {
                    startX: drag.startX,
                    startY: drag.startY,
                    currentX: event.pageX - this.containerOffset.left,
                    currentY: event.pageY - this.containerOffset.top,
                    selected: getCollisions(this.props.nodes, { left, top, right, bottom })
                }
            });
        }

        if (this.props.editorState.dragNodeUUID) {
            const startPosition = this.props.editorState.dragActive
                ? this.props.editorState.dragSelection.selected[this.props.editorState.dragNodeUUID]
                : this.props.nodes[this.props.editorState.dragNodeUUID].ui.position;

            const xd =
                event.pageX -
                this.containerOffset.left -
                this.props.editorState.dragDownPosition.left -
                startPosition.left;

            const yd =
                event.pageY -
                this.containerOffset.top -
                this.props.editorState.dragDownPosition.top -
                startPosition.top;

            if (this.props.editorState.dragActive) {
                this.props.onDragSelection({ left: xd, top: yd });
            } else {
                if (Math.abs(xd) + Math.abs(yd) > DRAG_THRESHOLD) {
                    const renderNode = this.props.nodes[this.props.editorState.dragNodeUUID];
                    const selection = this.props.editorState.dragSelection || {};
                    let selected = selection.selected || {};
                    if (!(renderNode.node.uuid in selected)) {
                        selected = { [renderNode.node.uuid]: renderNode.ui.position };
                    }

                    this.props.mergeEditorState({
                        dragActive: true,
                        dragSelection: {
                            selected
                        }
                    });
                }
            }
        }
    }

    public onMouseUp(event: React.MouseEvent<HTMLDivElement>): void {
        if (this.props.editorState.dragNodeUUID) {
            this.props.mergeEditorState({
                dragNodeUUID: null
            });
        }

        if (this.props.editorState.dragSelection && this.props.editorState.dragSelection.startX) {
            this.props.mergeEditorState({
                dragSelection: {
                    startX: null,
                    startY: null,
                    currentX: null,
                    currentY: null,
                    selected: this.props.editorState.dragSelection.selected
                }
            });

            // if (this.props.editorState.dragSelection.selected) {
            // this.Plumber.setDragSelection(this.props.editorState.dragSelection.selected);
            // }
        }
    }

    public getDragSelectionBox(): JSX.Element {
        if (this.props.editorState.dragSelection && this.props.editorState.dragSelection.startX) {
            return (
                <div
                    data-spec={dragSelectSpecId}
                    className={styles.dragSelection}
                    style={{ ...getDragStyle(this.props.editorState.dragSelection) }}
                />
            );
        }
        return null;
    }

    private onDragNodeStart(renderNode: RenderNode, position: FlowPosition): void {
        this.props.mergeEditorState({
            dragNodeUUID: renderNode.node.uuid,
            dragDownPosition: position
        });
    }

    private onDragNodeDrag(e: React.MouseEvent<HTMLDivElement>): void {}

    private onDragNodeStop(uuid: string): void {
        console.log('stopping:', uuid);
        // this.setState({ thisNodeDragging: false });
        this.props.mergeEditorState({
            dragNodeUUID: null,
            dragDownPosition: null,
            dragSelection: null,
            dragActive: false
        });
    }

    public render(): JSX.Element {
        return (
            <>
                {this.getSimulator()}
                {this.getDragNode()}
                {this.getNodeEditor()}
                <div
                    ref={this.onRef}
                    id={this.nodeContainerUUID}
                    className={styles.nodeList}
                    data-spec={nodesContainerSpecId}
                    onMouseDown={this.onMouseDown}
                    onMouseMove={this.onMouseMove}
                    onMouseUp={this.onMouseUp}
                    onDoubleClick={this.onDoubleClick}
                >
                    {this.getStickies()}
                    {this.getDragSelectionBox()}
                    {this.getNodes()}
                </div>
            </>
        );
    }
}

/* istanbul ignore next */
const mapStateToProps = ({
    flowContext: { definition, dependencies, nodes },
    editorState,
    nodeEditor: { settings }
}: AppState) => {
    return {
        nodeEditorSettings: settings,
        definition,
        nodes,
        dependencies,
        editorState: editorState as Partial<EditorState>
    };
};

/* istanbul ignore next */
const mapDispatchToProps = (dispatch: DispatchWithState) =>
    bindActionCreators(
        {
            mergeEditorState,
            ensureStartNode,
            resetNodeEditingState,
            onConnectionDrag,
            onOpenNodeEditor,
            onDragSelection,
            updateConnection,
            updateSticky
        },
        dispatch
    );

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Flow);
