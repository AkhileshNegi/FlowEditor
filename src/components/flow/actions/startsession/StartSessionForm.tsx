import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';
import Dialog, { ButtonSet } from '~/components/dialog/Dialog';
import { initializeForm, stateToAction } from '~/components/flow/actions/startsession/helpers';
import { ActionFormProps } from '~/components/flow/props';
import SelectAssetElement from '~/components/form/select/assets/SelectAssetElement';
import OmniboxElement from '~/components/form/select/omnibox/OmniboxElement';
import TypeList from '~/components/nodeeditor/TypeList';
import { fakePropType } from '~/config/ConfigProvider';
import { Asset } from '~/services/AssetService';
import { AssetArrayEntry, AssetEntry, FormState, mergeForm } from '~/store/nodeEditor';
import { validate, validateRequired } from '~/store/validators';

export interface StartSessionFormState extends FormState {
    recipients: AssetArrayEntry;
    flow: AssetEntry;
}

export default class StartSessionForm extends React.Component<
    ActionFormProps,
    StartSessionFormState
> {
    public static contextTypes = {
        endpoints: fakePropType,
        assetService: fakePropType
    };

    constructor(props: ActionFormProps) {
        super(props);

        this.state = initializeForm(this.props.nodeSettings);

        bindCallbacks(this, {
            include: [/^on/, /^handle/]
        });
    }

    public handleRecipientsChanged(recipients: Asset[]): boolean {
        return this.handleUpdate({ recipients });
    }

    public handleFlowChanged(flows: Asset[]): boolean {
        let flow = null;
        if (flows && flows.length > 0) {
            flow = flows[0];
        }
        return this.handleUpdate({ flow });
    }

    private handleUpdate(keys: { flow?: Asset; recipients?: Asset[] }): boolean {
        const updates: Partial<StartSessionFormState> = {};

        if (keys.hasOwnProperty('recipients')) {
            updates.recipients = validate('Recipients', keys.recipients, [validateRequired]);
        }

        if (keys.hasOwnProperty('flow')) {
            updates.flow = validate('Flow', keys.flow, [validateRequired]);
        }

        const updated = mergeForm(this.state, updates);
        this.setState(updated);
        return updated.valid;
    }

    private handleSave(): void {
        // validate in case they never updated an empty field
        const valid = this.handleUpdate({
            recipients: this.state.recipients.value,
            flow: this.state.flow.value
        });

        if (valid) {
            this.props.updateAction(stateToAction(this.props.nodeSettings, this.state));

            // notify our modal we are done
            this.props.onClose(false);
        }
    }

    private getButtons(): ButtonSet {
        return {
            primary: { name: 'Ok', onClick: this.handleSave },
            secondary: { name: 'Cancel', onClick: () => this.props.onClose(true) }
        };
    }

    public render(): JSX.Element {
        const typeConfig = this.props.typeConfig;
        return (
            <Dialog
                title={typeConfig.name}
                headerClass={typeConfig.type}
                buttons={this.getButtons()}
            >
                <TypeList
                    __className=""
                    initialType={typeConfig}
                    onChange={this.props.onTypeChange}
                />
                <div>
                    <OmniboxElement
                        data-spec="recipients"
                        name="Recipients"
                        assets={this.context.assetService.getRecipients()}
                        onChange={this.handleRecipientsChanged}
                        entry={this.state.recipients}
                        add={true}
                    />
                    <p>
                        <SelectAssetElement
                            key="flow_select"
                            name="Flow"
                            searchable={true}
                            placeholder="Select the flow to start them in"
                            entry={this.state.flow}
                            assets={this.context.assetService.getFlowAssets()}
                            onChange={this.handleFlowChanged}
                        />
                    </p>
                </div>
            </Dialog>
        );
    }
}
