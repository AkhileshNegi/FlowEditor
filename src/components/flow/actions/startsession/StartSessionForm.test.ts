import StartSessionForm from "components/flow/actions/startsession/StartSessionForm";
import { ActionFormProps } from "components/flow/props";
import { AssetType } from "store/flowContext";
import { composeComponentTestUtils, mock } from "testUtils";
import {
  createStartSessionAction,
  getActionFormProps,
  SubscribersGroup
} from "testUtils/assetCreators";
import * as utils from "utils";

mock(utils, "createUUID", utils.seededUUIDs());

const { setup } = composeComponentTestUtils<ActionFormProps>(
  StartSessionForm,
  getActionFormProps(createStartSessionAction())
);

describe(StartSessionForm.name, () => {
  describe("render", () => {
    it("should render self, children with base props", () => {
      const { wrapper } = setup(true);
      expect(wrapper).toMatchSnapshot();
    });

    it("should render an empty form with no action", () => {
      const { wrapper, instance } = setup(true, {
        $merge: {
          nodeSettings: { originalNode: null, originalAction: null }
        }
      });

      expect(instance.state).toMatchSnapshot();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("updates", () => {
    it("should save changes", () => {
      const { instance, props } = setup(true);

      instance.handleRecipientsChanged([SubscribersGroup]);
      instance.handleFlowChanged([
        { id: "my_flow", name: "My Flow", type: AssetType.Flow }
      ]);
      expect(instance.state).toMatchSnapshot();

      instance.handleSave();
      expect(props.updateAction).toHaveBeenCalled();
      expect(props.updateAction).toMatchCallSnapshot();
    });

    it("should allow switching from router", () => {
      const { instance, props } = setup(true, {
        $merge: { updateAction: jest.fn() },
        nodeSettings: { $merge: { originalAction: null } }
      });

      instance.handleRecipientsChanged([SubscribersGroup]);
      instance.handleFlowChanged([
        { id: "my_flow", name: "My Flow", type: AssetType.Flow }
      ]);
      instance.handleSave();

      expect(props.updateAction).toMatchCallSnapshot();
    });
  });

  describe("cancel", () => {
    it("should cancel without changes", () => {
      const { instance, props } = setup(true, {
        $merge: { onClose: jest.fn(), updateAction: jest.fn() }
      });

      instance.handleRecipientsChanged([SubscribersGroup]);
      instance.handleFlowChanged([
        { id: "my_flow", name: "My Flow", type: AssetType.Flow }
      ]);
      instance.getButtons().secondary.onClick();
      expect(props.onClose).toHaveBeenCalled();
      expect(props.updateAction).not.toHaveBeenCalled();
    });
  });
});
