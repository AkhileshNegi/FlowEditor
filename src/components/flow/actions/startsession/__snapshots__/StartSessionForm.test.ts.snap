// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StartSessionForm render should render an empty form with no action 1`] = `
Object {
  "flow": Object {
    "value": null,
  },
  "recipients": Object {
    "value": Array [],
  },
  "valid": false,
}
`;

exports[`StartSessionForm render should render an empty form with no action 2`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="start_session"
  title="Start Somebody Else"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "component": [Function],
        "description": "Start somebody else in a flow",
        "form": undefined,
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
        ],
        "name": "Start Somebody Else",
        "type": "start_session",
      }
    }
    onChange={[MockFunction]}
  />
  <div>
    <OmniboxElement
      add={true}
      assets={
        RecipientAssets {
          "assets": Object {},
          "endpoint": "/assets/recipients.json",
          "idProperty": "uuid",
          "localStorage": true,
        }
      }
      data-spec="recipients"
      entry={
        Object {
          "value": Array [],
        }
      }
      name="Recipients"
      onChange={[Function]}
      placeholder="Enter a group or contact..."
      searchPromptText="No matches"
    />
    <p>
      <SelectAssetElement
        assets={
          FlowAssets {
            "assetType": "flow",
            "assets": Object {},
            "endpoint": "/assets/flows.json",
            "idProperty": "uuid",
            "localStorage": true,
          }
        }
        entry={
          Object {
            "value": null,
          }
        }
        key="flow_select"
        name="Flow"
        onChange={[Function]}
        placeholder="Select the flow to start them in"
        searchable={true}
      />
    </p>
  </div>
</Dialog>
`;

exports[`StartSessionForm render should render self, children with base props 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="start_session"
  title="Start Somebody Else"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "component": [Function],
        "description": "Start somebody else in a flow",
        "form": undefined,
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
        ],
        "name": "Start Somebody Else",
        "type": "start_session",
      }
    }
    onChange={[MockFunction]}
  />
  <div>
    <OmniboxElement
      add={true}
      assets={
        RecipientAssets {
          "assets": Object {},
          "endpoint": "/assets/recipients.json",
          "idProperty": "uuid",
          "localStorage": true,
        }
      }
      data-spec="recipients"
      entry={
        Object {
          "value": Array [
            Object {
              "id": "4441fa19-8bbf-4894-8529-d04cc3b365d6",
              "name": "Cat Fanciers",
              "type": "group",
            },
            Object {
              "id": "1e5aa7db-89c7-4dec-99b8-cd1194e1f46e",
              "name": "Cat Facts",
              "type": "group",
            },
            Object {
              "id": "1e5aa7db-89c7-4dec-99b8-cd1194e1f46e",
              "name": "Kellan Alexander",
              "type": "contact",
            },
            Object {
              "id": "575fe8e9-cb51-4f15-8df7-422290fdfc64",
              "name": "Norbert Kwizera",
              "type": "contact",
            },
            Object {
              "id": "82df72ae-d835-401b-b248-d0c5dcfdce5c",
              "name": "Rowan Seymour",
              "type": "contact",
            },
          ],
        }
      }
      name="Recipients"
      onChange={[Function]}
      placeholder="Enter a group or contact..."
      searchPromptText="No matches"
    />
    <p>
      <SelectAssetElement
        assets={
          FlowAssets {
            "assetType": "flow",
            "assets": Object {},
            "endpoint": "/assets/flows.json",
            "idProperty": "uuid",
            "localStorage": true,
          }
        }
        entry={
          Object {
            "value": Object {
              "id": "flow_uuid",
              "name": "Flow to Start",
              "type": "flow",
            },
          }
        }
        key="flow_select"
        name="Flow"
        onChange={[Function]}
        placeholder="Select the flow to start them in"
        searchable={true}
      />
    </p>
  </div>
</Dialog>
`;

exports[`StartSessionForm updates should allow switching from router 1`] = `
Array [
  Object {
    "contacts": Array [],
    "flow": Object {
      "name": "My Flow",
      "uuid": "my_flow",
    },
    "groups": Array [
      Object {
        "name": "Subscriber",
        "uuid": "68223118-109f-442a-aed3-7bb3e1eab687",
      },
    ],
    "type": "start_session",
    "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
  },
]
`;

exports[`StartSessionForm updates should save changes 1`] = `
Object {
  "flow": Object {
    "value": Object {
      "id": "my_flow",
      "name": "My Flow",
      "type": "flow",
    },
  },
  "recipients": Object {
    "value": Array [
      Object {
        "id": "68223118-109f-442a-aed3-7bb3e1eab687",
        "name": "Subscriber",
        "type": "group",
      },
    ],
  },
  "valid": true,
}
`;

exports[`StartSessionForm updates should save changes 2`] = `
Array [
  Object {
    "contacts": Array [],
    "flow": Object {
      "name": "My Flow",
      "uuid": "my_flow",
    },
    "groups": Array [
      Object {
        "name": "Subscriber",
        "uuid": "68223118-109f-442a-aed3-7bb3e1eab687",
      },
    ],
    "type": "start_session",
    "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
  },
]
`;
